proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1232 @@
+}
+extension Character.UTF8View: ContiguousStorage<Unicode.UTF8.CodeUnit> {
+  // note: this could borrow a temporary copy of the `Character`'s storage
+  var storage: Span<Unicode.UTF8.CodeUnit> { get }
+}
+
+extension SIMD: ContiguousStorage<Self.Scalar> {
+  var storage: Span<Self.Scalar> { get }
+}
+extension KeyValuePairs: ContiguousStorage<(Self.Key, Self.Value)> {
+  var storage: Span<(Self.Key, Self.Value)> { get }
+}
+extension CollectionOfOne: ContiguousStorage<Element> {
+  var storage: Span<Element> { get }
+}
+
+extension Slice: ContiguousStorage where Base: ContiguousStorage {
+  var storage: Span<Base.Element> { get }
+}
+```
+
+In addition to the the safe types above gaining the `storage` property, the `UnsafeBufferPointer` family of types will also gain access to a `storage` property. This enables interoperability if `Span`-taking API. While a `Span` binding created from an `UnsafeBufferPointer` exists, the memory that underlies it must not be deinitialized or deallocated.
+
+```swift
+extension UnsafeBufferPointer: ContiguousStorage<Self.Element> {
+  // note: additional preconditions apply until the end of the scope
+  var storage: Span<Element> { get }
+}
+extension UnsafeMutableBufferPointer: ContiguousStorage<Self.Element> {
+  // note: additional preconditions apply until the end of the scope
+  var storage: Span<Element> { get }
+}
+extension UnsafeRawBufferPointer: ContiguousStorage<UInt8> {
+  // note: additional preconditions apply until the end of the scope
+  var storage: Span<UInt8> { get }
+}
+extension UnsafeMutableRawBufferPointer: ContiguousStorage<UInt8> {
+  // note: additional preconditions apply until the end of the scope
+  var storage: Span<UInt8> { get }
+}
+```
+
+#### Using `Span` with C functions or other unsafe code:
+
+`Span` has an unsafe hatch for use with unsafe code.
+
+```swift
+extension Span {
+  func withUnsafeBufferPointer<Result>(
+    _ body: (_ buffer: UnsafeBufferPointer<Element>) -> Result
+  ) -> Result
+}
+
+extension Span where Element: BitwiseCopyable {
+  func withUnsafeBytes<Result>(
+    _ body: (_ buffer: UnsafeRawBufferPointer) -> Result
+  ) -> Result
+}
+```
+
+
+
+### Complete `Span` API:
+
+```swift
+public struct Span<Element: ~Copyable & ~Escapable>: Copyable, ~Escapable {
+  internal var _start: Span<Element>.Index
+  internal var _count: Int
+}
+```
+
+##### Creating a `Span`:
+
+The initialization of a `Span` instance is an unsafe operation. When it is initialized correctly, subsequent uses of the borrowed instance are safe. Typically these initializers will be used internally to a container's implementation of functions or computed properties that return a borrowed `Span`.
+
+```swift
+extension Span {
+
+  /// Unsafely create a `Span` over initialized memory.
+  ///
+  /// The memory in `buffer` must be owned by the instance `owner`,
+  /// meaning that as long as `owner` is alive the memory will remain valid.
+  ///
```

Comment:
Returns `nil` if `buffer.baseAddress` is `nil`.

Reply:
@lorentey would prefer this to not be failable, fwiw.

Reply:
We will make this non-failable.

Reply:
Would it trap on nil or make an empty span on nil (and would base address he 0?)?

Reply:
It would need to conjur a pointer value out of thin air. We can define an empty singleton for this purpose (say, a dummy UInt8), and put it in a read-only section.

Reply:
(Trapping and returning nil both have the fatal flaw that they force the client to care about whether their empty buffer happens to have a nil base. That would be terrible -- the client isn't necessarily in a position to dictate that, and forcing them to try to somehow manually handle the nil case would only lead to entirely avoidable bugs.)

Reply:
There are no failable initializers anymore in the proposal. The implementation still traps when `baseAddress` is `nil`; I think we can defer conjuring a pointer until adapting the work in the standard library.